### PR TITLE
ConvertTokenizer, CLS token & features as sequence 

### PR DESCRIPTION
**Pull request by tabergma**
**Proposed changes**:
- Add ConveRT Tokenizer
- Remove option `use_cls_token` from all tokenizers. Tokenizers add CLS token by default now.
- Remove option `return_sequence` from featurizers. Featurizers return sequence by default. The feature vector of the CLS token contains the feature for the complete utterance. 
- Implement `train` and `process` of tokenizers in `Tokenizer` class. Subclasses just need to implement `tokenize`.

closes https://github.com/RasaHQ/rasa/issues/4978

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 1]
```diff
@@ -54,7 +56,7 @@ class EmbeddingIntentClassifier(Component):
 
     provides = ["intent", "intent_ranking"]
 
-    requires = []
+    requires = [TOKENS_NAMES[TEXT_ATTRIBUTE]]
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]: maybe leave a comment why it requires tokens

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 2]
```diff
@@ -526,9 +525,10 @@ def _combine_sparse_dense_features(
             else:
                 dense_features.append(f)
 
-        output = tf.concat(dense_features, axis=-1) * mask
-        # apply mean to convert sequence to sentence features
-        output = tf.reduce_sum(output, axis=1) / tf.reduce_sum(mask, axis=1)
+        # extract the CLS token feature vector
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]: let's extract the `cls` token before applying `dense_layer_for_sparse`, then we save the computation over the entire sequence

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 3]
```diff
@@ -526,9 +525,10 @@ def _combine_sparse_dense_features(
...
+        dense_features = [f[:, -1, :] for f in dense_features]
```
Comment (Ghostvv) [inline rasa/nlu/classifiers/embedding_intent_classifier.py]: I think we need a mask to extract the last token due to different sequence lengths. Let's use the following:
```
sequence_lengths = tf.cast(tf.reduce_sum(mask[:, :, 0], 1), tf.int32)
last_index = tf.maximum(tf.constant(0, dtype=sequence_lengths.dtype), sequence_lengths - 1)
idxs = tf.stack([tf.range(tf.shape(last_index)[0]), last_index], axis=1)
cls = tf.gather_nd(text_transformed, idxs)
```

File: rasa/nlu/extractors/crf_entity_extractor.py [hunk 4]
```diff
@@ -615,7 +615,7 @@ def _from_text_to_crf(
         if self.pos_features:
             tokens = message.get(SPACY_DOCS[TEXT_ATTRIBUTE])
         else:
-            tokens = message.get(TOKENS_NAMES[TEXT_ATTRIBUTE])
+            tokens = message.get(TOKENS_NAMES[TEXT_ATTRIBUTE])[:-1]
```
Comment (Ghostvv) [inline rasa/nlu/extractors/crf_entity_extractor.py]: maybe add a comment that we're removing `cls` token

File: rasa/nlu/extractors/mitie_entity_extractor.py [hunk 5]
```diff
@@ -89,16 +89,19 @@ def train(
...
             self.ner = trainer.train()
 
-    def _prepare_mitie_sample(self, training_example) -> Any:
+    def _prepare_mitie_sample(self, training_example: Message) -> Any:
         import mitie
 
         text = training_example.text
         tokens = training_example.get(TOKENS_NAMES[TEXT_ATTRIBUTE])
-        sample = mitie.ner_training_instance([t.text for t in tokens])
+        tokens_without_cls = tokens[:-1]
```
Comment (Ghostvv) [inline rasa/nlu/extractors/mitie_entity_extractor.py]: in crf extractor, we add `[:-1]` when we read tokens, I guess it makes sense to do the same for consistency, with a comment

File: rasa/nlu/extractors/mitie_entity_extractor.py [hunk 6]
```diff
@@ -89,16 +89,19 @@ def train(
         if found_one_entity:
             self.ner = trainer.train()
 
-    def _prepare_mitie_sample(self, training_example) -> Any:
+    def _prepare_mitie_sample(self, training_example: Message) -> Any:
         import mitie
 
         text = training_example.text
         tokens = training_example.get(TOKENS_NAMES[TEXT_ATTRIBUTE])
```
Comment (Ghostvv) [inline rasa/nlu/extractors/mitie_entity_extractor.py]:
```suggestion
        tokens = training_example.get(TOKENS_NAMES[TEXT_ATTRIBUTE])[:-1]
```

File: rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py [hunk 7]
```diff
@@ -328,9 +322,11 @@ def _convert_attribute_tokens_to_texts(
         attribute_texts = {}
 
         for attribute in attribute_tokens.keys():
-            attribute_texts[attribute] = [
-                " ".join(tokens) for tokens in attribute_tokens[attribute]
-            ]
+            tokens = attribute_tokens[attribute]
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py]: let's call it `all_tokens` or smth, because it is a list of a list of tokens
```suggestion
            all_tokens = attribute_tokens[attribute]
```

File: rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py [hunk 8]
```diff
@@ -328,9 +322,11 @@ def _convert_attribute_tokens_to_texts(
...
+            if attribute in [RESPONSE_ATTRIBUTE, TEXT_ATTRIBUTE]:
+                # vocabulary should not contain CLS token
+                tokens = [t[:-1] for t in tokens]
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py]:
```suggestion
                all_tokens = [tokens[:-1] for tokens in all_tokens]
```

File: rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py [hunk 9]
```diff
@@ -328,9 +322,11 @@ def _convert_attribute_tokens_to_texts(
...
+            attribute_texts[attribute] = [" ".join(t) for t in tokens]
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py]:
```suggestion
            attribute_texts[attribute] = [" ".join(tokens) for tokens in all_tokens]
```

File: rasa/nlu/featurizers/sparse_featurizer/regex_featurizer.py [hunk 10]
```diff
@@ -112,15 +101,17 @@ def _features_for_patterns(
                 patterns = t.get("pattern", default={})
                 patterns[pattern["name"]] = False
 
-                if self.return_sequence:
-                    seq_index = token_index
-                else:
-                    seq_index = 0
+                if t.text == CLS_TOKEN:
+                    t.set("pattern", patterns)
+                    continue
```
Comment (Ghostvv) [inline rasa/nlu/featurizers/sparse_featurizer/regex_featurizer.py]: could you please comment

File: rasa/nlu/tokenizers/convert_tokenizer.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+import tensorflow as tf
+
+
+class ConveRTTokenizer(WhitespaceTokenizer):
+
+    provides = [TOKENS_NAMES[attribute] for attribute in MESSAGE_ATTRIBUTES]
+
+    defaults = {
+        # Flag to check whether to split intents
+        "intent_tokenization_flag": False,
```
Comment (Ghostvv) [inline rasa/nlu/tokenizers/convert_tokenizer.py]: I think other tokenizersdon't provide this functionality, and users would need to add whitespace one for spacy. I think, it makes sense to silent this behavior here

File: rasa/nlu/tokenizers/convert_tokenizer.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        message.set(
+            TOKENS_NAMES[TEXT_ATTRIBUTE], self.tokenize_using_convert(message.text)
+        )
+
+    def _tokenize(self, sentence: Text) -> Any:
+        return self.session.run(
+            self.tokenized, feed_dict={self.text_placeholder: [sentence]}
+        )
+
+    def tokenize_using_convert(
```
Comment (Ghostvv) [inline rasa/nlu/tokenizers/convert_tokenizer.py]: shall we override tokenize function?

File: rasa/nlu/tokenizers/convert_tokenizer.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+
+        ConveRT adds a special char in front of (some) words and splits words into
+        sub-words. To ensure the entity start and end values matches the token values,
+        tokenize the text first using the whitespace tokenizer. If individual tokens
+        are split up into multiple tokens, we make sure that the start end end value
+        of the first and last respective tokens stay the same.
+        """
+
+        # perform whitespace tokenization
+        tokens_in = self.tokenize(text, attribute)
```
Comment (Ghostvv) [inline rasa/nlu/tokenizers/convert_tokenizer.py]: and here call `super.tokenize`

File: rasa/nlu/tokenizers/convert_tokenizer.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+        # remove CLS token if present
+        if tokens_in[-1].text == CLS_TOKEN:
```
Comment (Ghostvv) [inline rasa/nlu/tokenizers/convert_tokenizer.py]: in different algorithms we either check ` if attribute in [RESPONSE_ATTRIBUTE, TEXT_ATTRIBUTE]:` or `if tokens_in[-1].text == CLS_TOKEN:` I guess we should use the same if statement everywhere if possible

File: rasa/nlu/tokenizers/convert_tokenizer.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+                split_token_strings, token_end, token_start
+            )
+            tokens_out += _aligned_tokens
+
+        tokens_out = self.add_cls_token(tokens_out, attribute)
+
+        return tokens_out
+
+    def _clean_tokens(self, tokens: List[Text]):
+        """Encode tokens and remove special char added by ConveRT."""
```
Comment (Ghostvv) [inline rasa/nlu/tokenizers/convert_tokenizer.py]:
```suggestion
        """Encode tokens and remove special char added by ConveRT."""
        
```

File: rasa/nlu/tokenizers/convert_tokenizer.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+        tokens = [string.decode("utf-8").replace("﹏", "") for string in tokens]
+        return [string for string in tokens if string]
+
+    def _align_tokens(self, tokens_in: List[Text], token_end: int, token_start: int):
```
Comment (Ghostvv) [inline rasa/nlu/tokenizers/convert_tokenizer.py]: please add comments, what do we do here

File: rasa/nlu/classifiers/embedding_intent_classifier.py (hunk 3 above)
Comment (tabergma) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: true, but should not be needed anymore once we extract the features already in `create_session_data`

Comment (Ghostvv) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: oh, then yes, I didn't see that

**Comment by tabergma** [comment]
@Ghostvv suggested to move the `train` method of the tokenizers to the superclass and let the tokenizers just overwrite the `tokenize` method. This way we can ensure that the CLS token is always added (we can do it inside `train`). We should check this idea.

File: rasa/nlu/tokenizers/convert_tokenizer.py (hunk 11 above)
Comment (tabergma) [reply rasa/nlu/tokenizers/convert_tokenizer.py]: the whitespace tokenizer has the same options

File: rasa/nlu/classifiers/embedding_intent_classifier.py [hunk 17]
```diff
@@ -531,8 +537,8 @@ def _combine_sparse_dense_features(
                 dense_features.append(f)
 
         output = tf.concat(dense_features, axis=-1) * mask
-        # apply mean to convert sequence to sentence features
-        output = tf.reduce_sum(output, axis=1) / tf.reduce_sum(mask, axis=1)
+        output = tf.reduce_sum(output, axis=1)
```
[the remark this answers is not in the document]
Comment (Ghostvv) [reply rasa/nlu/classifiers/embedding_intent_classifier.py]: let's substitute it with `tf.squeeze` so we don't forget why it is here

File: changelog/4978.removal.rst [hunk 18]
```diff
@@ -0,0 +1,9 @@
+Remove option ``use_cls_token`` from tokenizers and option ``return_sequence`` from featurizers.
+
+By default all tokenizer add a special token (``__CLS__``) to the end of the list of tokens.
+This token will be used to capture the features of the whole utterance.
+
+The featurizers will return a matrix of size (token-length x feature-dimension) by default.
```
Comment (dakshvar22) [inline changelog/4978.removal.rst]:
```suggestion
The featurizers will return a matrix of size (number-of-tokens x feature-dimension) by default.
```
I got confused initially if by `token-length` you meant length of an individual token. Suggested this for clarity.

File: docs/api/custom-nlu-components.rst [hunk 19]
```diff
@@ -36,25 +36,15 @@ methods that you should implement:
...
-    If you create a custom featurizer you should have the following entry in your default dict.
-
-    .. code-block:: none
-
-        # if True return a sequence of features (return vector has size
-        # token-size x feature-dimension)
-        # if False token-size will be equal to 1
-        "return_sequence": False,
+    If you create a custom featurizer you should return a sequence of features.
+    E.g. your featurizer should return a matrix of size (token-size x feature-dimension).
```
Comment (dakshvar22) [inline docs/api/custom-nlu-components.rst]:
```suggestion
    E.g. your featurizer should return a matrix of size (number-of-tokens x feature-dimension).
```

File: docs/nlu/components.rst [hunk 20]
```diff
@@ -91,15 +91,12 @@ As those feature vectors would normally take up a lot of memory, we store them a
...
-By default all featurizers will return a matrix of length (1 x feature-dimension).
-All featurizer (except the ``ConveRTFeaturizer``) have the option to return a sequence instead.
-In case the flag ``"return_sequence"`` is set to ``True``, the returned matrix of the featurizer will have the size
-(token-length x feature-dimension).
-So, the returned matrix will have an entry for every token.
-Otherwise, the matrix will just have one entry for the complete utterance.
-If you want to use custom features for your ``CRFEntityExtractor``, you should set ``"return_sequence"`` to ``True``.
-For more details, take a look :ref:`entity-extraction-custom-features`.
-
+By default all featurizers will return a matrix of length (token-length x feature-dimension).
```
Comment (dakshvar22) [inline docs/nlu/components.rst]:
```suggestion
By default all featurizers will return a matrix of length (number-of-tokens x feature-dimension).
```

File: rasa/nlu/featurizers/dense_featurizer/convert_featurizer.py [hunk 21]
```diff
@@ -72,19 +61,93 @@ def _compute_features(
...
         # Get text for attribute of each example
         batch_attribute_text = [ex.get(attribute) for ex in batch_examples]
+        sentence_encodings = self._sentence_encoding_of_text(batch_attribute_text)
 
-        batch_features = self._run_model_on_text(batch_attribute_text)
+        # convert them to a sequence of 1
+        return np.reshape(sentence_encodings, (len(batch_examples), 1, -1))
 
-        return batch_features
+    def _compute_sequence_encodings(
```
Comment (dakshvar22) [inline rasa/nlu/featurizers/dense_featurizer/convert_featurizer.py]: Optional suggestion: 
Add the pre-computed sentence encodings to the list of sequence encodings outside of this function since the name of the function doesn't suggest to do that.

File: rasa/nlu/tokenizers/whitespace_tokenizer.py [hunk 22]
```diff
@@ -23,76 +17,39 @@ class WhitespaceTokenizer(Tokenizer):
...
+            r"(\s|^)[^\w#@&]+(?=[^0-9\s])|"
+            # not in between numbers and not . or @ or & or - or #
+            # e.g. 10'000.00 or blabla@gmail.com
+            # and not url characters
+            r"(?<=[^0-9\s])[^\w._~:/?#\[\]()@!$&*+,;=-]+(?=[^0-9\s])",
+            " ",
+            text,
+        ).split()
+        # Fix for convert tokenizer, remove any single '&'
+        words = [w for w in words if w != "&"]
```
Comment (dakshvar22) [inline rasa/nlu/tokenizers/whitespace_tokenizer.py]: Do we still need this?

Comment (tabergma) [reply rasa/nlu/tokenizers/whitespace_tokenizer.py]: I guess, we will get a "dimensions do not match" kind of error if we remove it. Can quickly test it.

Comment (dakshvar22) [reply rasa/nlu/tokenizers/whitespace_tokenizer.py]: If we use only one tokenizer which would be now ConveRT tokenizer, we shouldn't get it I guess. Earlier the problem was that the method of tokenization was different between ConveRTTokenizer and ConveRTFeaturizer. Now that we have it covered in one place only, I think it should be fine. Can you do a quick test please?

Comment (tabergma) [reply rasa/nlu/tokenizers/whitespace_tokenizer.py]: Yes, tested it. We do not need it anymore. Will remove it.
